### PR TITLE
mon: add ability to mute health alerts

### PR DESCRIPTION
**Pull request by liewegas**
TODO
- [x] add qa/standalone test(s)
- [x] documentation mute/unmute commands, behavior

### CODE REVIEW

**Comment by liewegas** [comment]
retest this please

**Comment by sebastian-philipp** [comment]
@LenzGr do we need a dashboard integration?

**Comment by liewegas** [comment]
> @LenzGr do we need a dashboard integration?

Definitely!

File: qa/standalone/mon/health-mute.sh [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    ceph health detail | grep OSDMAP_FLAGS || return 1
+    ceph osd down 0
+    ceph -s
+    ceph health detail | grep OSD_DOWN || return 1
+    ceph health detail | grep HEALTH_WARN || return 1
+
+    ceph health mute OSD_DOWN
+    ceph health mute OSDMAP_FLAGS
+    ceph -s
+    ceph health | grep HEALTH_OK | grep OSD_DOWN | grep OSDMAP_FLAGS || return 1
```
Comment (dzafman) [inline qa/standalone/mon/health-mute.sh]: You need "set -o pipefail" if you want this to fail if any of the 3 greps fail.

File: qa/standalone/mon/health-mute.sh [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    ceph health unmute OSD_DOWN
+    ceph health | grep HEALTH_WARN
```
Comment (dzafman) [inline qa/standalone/mon/health-mute.sh]: Add || return 1

File: src/mon/HealthMonitor.cc [hunk 3]
```diff
@@ -161,11 +196,154 @@ bool HealthMonitor::prepare_update(MonOpRequestRef op)
...
+  string format;
+  cmd_getval(g_ceph_context, cmdmap, "format", format);
+  boost::scoped_ptr<Formatter> f(Formatter::create(format, "json-pretty",
+						   "json-pretty"));
+
+  string prefix;
+  cmd_getval(g_ceph_context, cmdmap, "prefix", prefix);
+  int r = 0;
+
+//} else {
```
Comment (dzafman) [inline src/mon/HealthMonitor.cc]: ???

Comment (liewegas) [reply src/mon/HealthMonitor.cc]: most of the method is boilerplate but there are no read-only commands here yet.  could take it all out i guess but didn't realize at the time we wouldn't implement any

Comment (liewegas) [reply src/mon/HealthMonitor.cc]: actually, i should probably most the "health" implementation from Monitor.cc to here

File: doc/rados/operations/health-checks.rst [hunk 4]
```diff
@@ -71,10 +72,78 @@ listen for v2 connections on the new default 3300 port.
...
+  ceph config set global mon_data_size_warn <size>
+
 
 Manager
 -------
 
+MGR_DOWN
+________
+
+All manager` daemons are currently down.  The cluster should normally
```
Comment (neha-ojha) [inline doc/rados/operations/health-checks.rst]: nit: extra "`"

File: doc/rados/operations/health-checks.rst [hunk 5]
```diff
@@ -71,10 +72,78 @@ listen for v2 connections on the new default 3300 port.
...
+have at least one running manager (``ceph-mgr``) daemon.  If no
+manager daemon is running, the clusters ability to monitor itself will
```
Comment (neha-ojha) [inline doc/rados/operations/health-checks.rst]: nit: cluster's

File: doc/rados/operations/health-checks.rst [hunk 6]
```diff
@@ -891,6 +891,28 @@ You can manually initiate a scrub of a clean PG with::
...
+configured warning threshold.  This indicates that either an extremely
+large number of snapshots were recently deleted, or that the OSDs are
+unable to trim snapshots quickly enough to keep up with the rate of
+new snapshot deletions.
+
+The warning threshold is controlled by the
+``mon_osd_snap_trim_queue_warn_on`` option (default: 32768).
+
+This warning may trigger if OSDs are under excessive load and unable
+to keep up with their background work, or if the OSDs internal
```
Comment (neha-ojha) [inline doc/rados/operations/health-checks.rst]: nit: OSD's or OSDs'

File: doc/rados/operations/health-checks.rst [hunk 7]
```diff
@@ -71,6 +71,56 @@ listen for v2 connections on the new default 3300 port.
...
+
+MON_DISK_BIG
+____________
+
+The database size for one or more monitors is very large.  This alert
+triggers if the size of the monitor's database is larger than
+``mon_data_size_warn`` (default: 15 GiB).
+
+A large database is unusual, but may not necessarily indicate a
+problem.  Monitor databases may grow in size of there are placement
```
Comment (neha-ojha) [inline doc/rados/operations/health-checks.rst]: s/of/when/

File: doc/rados/operations/monitoring.rst [hunk 8]
```diff
@@ -159,6 +159,62 @@ to a health state:
...
+A mute can be explicitly removed with::
+
+  ceph health unmute <code>
+
+For example,::
+
+  ceph health unmute OSD_DOWN
+
+A health check mute may optionally have a TTL (time to live)
+associated with it, sucht that the mute will automatically expire
```
Comment (neha-ojha) [inline doc/rados/operations/monitoring.rst]: s/sucht/such

File: src/mon/MonCommands.h [hunk 9]
```diff
@@ -211,6 +211,11 @@ COMMAND_WITH_FLAG("injectargs " \
 COMMAND("status", "show cluster status", "mon", "r")
 COMMAND("health name=detail,type=CephChoices,strings=detail,req=false", \
 	"show cluster health", "mon", "r")
+COMMAND("health mute name=code,type=CephString name=ttl,type=CephString,req=false " \
```
Comment (neha-ojha) [inline src/mon/MonCommands.h]: nit: having each name in a fresh line may be easier to read

**Comment by smithfarm** [comment]
@liewegas Since https://tracker.ceph.com/issues/40420 references this PR as its fix, I added that URL to the description.

Should this feature be backported to nautilus?

**Pull request merged**